internal/data/metric.go

```diff
@@ -65,42 +77,145 @@ func (md MetricData) MetricCount() int {
 	return metricCount
 }
 
+// ToOtlp converts the internal MetricData to the OTLP.
+func (md MetricData) ToOtlp() []*otlpmetrics.ResourceMetrics {
+	if !md.pimpl.initialized {
+		// Guaranteed no changes via internal fields.
+		return md.orig
+	}
+	if md.pimpl.sliceChanged {
+		// Reconstruct the slice because we don't know what elements were removed/added.
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		md.orig = make([]*otlpmetrics.ResourceMetrics, len(md.pimpl.resourceMetrics))
+		for i := range md.pimpl.resourceMetrics {
+			md.orig[i] = md.pimpl.resourceMetrics[i].getOrig()
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	} else {
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		for i := range md.pimpl.resourceMetrics {
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	}
+	return md.orig
+}
+
+func (md MetricData) initInternallIfNeeded() {
+	if !md.pimpl.initialized {
+		md.pimpl.resourceMetrics = newResourceMetricsSliceFromOrig(md.orig)
+		md.pimpl.initialized = true
+	}
+}
+
 // ResourceMetrics is a collection of metrics from a Resource.
 //
 // Must use NewResourceMetrics functions to create new instances.
 // Important: zero-initialized instance is not valid for use.
 type ResourceMetrics struct {
-	orig *otlpmetric.ResourceMetrics
+	orig *otlpmetrics.ResourceMetrics
 
 	// Override a few fields. These fields are the source of truth. Their counterparts
 	// stored in corresponding fields of "orig" are ignored.
 	pimpl *internalResourceMetrics
 }
 
 type internalResourceMetrics struct {
-	resource *Resource
+	resource Resource
 	metrics  []Metric
+	// True when the slice was replace.
+	sliceChanged bool
+	// True if the pimpl was initialized.
+	initialized bool
 }
 
 // NewResourceMetrics creates a new ResourceMetrics.
 func NewResourceMetrics() ResourceMetrics {
-	return ResourceMetrics{&otlpmetric.ResourceMetrics{}, &internalResourceMetrics{}}
+	return ResourceMetrics{&otlpmetrics.ResourceMetrics{}, &internalResourceMetrics{}}
```

Comment:
Another optimization could be declaring a struct like this:
```go
type resourceMetricsAlloc struct {
	orig otlpmetric.ResourceMetrics
	pimpl internalResourceMetrics
}
```
and using a single allocation instead of 2:
```go
a := &resourceMetricsAlloc{}
return ResourceMetrics{&a.orig, &a.pimpl}
```

Reply:
Added to the optimization issue.